README.md
*friday* is an image processing framework for Haskell. It has been designed to
build fast, generic and type-safe image processing algorithms.

*friday* also provide some simple computer vision features such as edge
detection or histogram processing.

Except for I/Os, *friday* is entirely written in Haskell.

![Header](header.png)

# Features

The four distinguishing features of this image library are:

* A reliance on a strongly typed programming language to detect programming
  errors directly at compile time.
* The ability to fuse image transformations.
  For example, if one wishes to apply a rotation on a resized image, he can
  use *friday* and the *Haskell* compiler to generate a single loop that will
  automatically combine the resizing and the rotating operations into a single
  algorithm, removing the need to store the intermediate resized image.
* The ability to automatically parallelize algorithms to use multiple
  processors.
  The library is able to parallelize upon request image transformations, even
  these which have been generated by the fusion mechanism.
* Being extremely generic. One who wants to create new algorithms, new
  pixel color-spaces or new ways to store an image will be able to reuse the 
  advanced type checking features and both the fusion and automatic
  parallelization mechanisms.

The library is more like a set of *building blocks to write image processing
algorithms in a functional programming style* than a collection of image
processing algorithms.

The library currently supports four color-spaces: RGB, RGBA, HSV and gray-scale
images. Images can be converted between these color-spaces.

At this moment, the following features and algorithms have been implemented:

* various image transformations: resizing, cropping, vertical and horizontal
[flop/flipping](http://en.wikipedia.org/wiki/Flopped_image) and
[flood fill](http://en.wikipedia.org/wiki/Flood_filling) ;

* filters:
[morphological transformations](http://en.wikipedia.org/wiki/Mathematical_morphology)
(dilation and erosion), blurring (mean
and [Gaussian](http://en.wikipedia.org/wiki/Gaussian_blur) blurs) and
derivative computation
([Sobel](http://en.wikipedia.org/wiki/Sobel_operator) and
[Scharr](http://en.wikipedia.org/wiki/Sobel_operator#Alternative_operators)
operators) ;

* support for mutable and masked images ;

* non-adaptive, adaptive, Otsu and SCW thresholding methods ;

* edge detection using
[Canny's algorithm](http://en.wikipedia.org/wiki/Canny_edge_detector) ;

* color histograms (including comparisons and image equalization).

# Quick tour

## Modules

The library is divided in three main modules: `Vision.Primitive`, `Vision.Image`
and `Vision.Histogram`.

You can directly import sub-modules (such as `Vision.Image.Transform`), but the
backward-compatibility is *not* enforced. That is, some functions could be moved
to another sub-module in a newer backward-compatible advertised version.

### Vision.Primitive

`Vision.Primitive` contains types used all over the library, such as `Shape`
which is used to define shapes and indices in images and histograms.

The module is usually imported unqualified:

```haskell
import Vision.Primitive
```

### Vision.Image

`Vision.Image` contains functions and types to manage and transform images.

The module can be imported unqualified but it's often better use a qualified
import as some function names (such as `map`) can conflict with `Prelude` and
`Vision.Histogram`:

```haskell
import Vision.Image
-- or
import qualified Vision.Image as I
```

### Vision.Histogram

`Vision.Histogram` contains functions and types to create and compare
histograms.

The module can be imported unqualified but it's often better use a qualified
import as some function names (such as `map`) can conflict with `Prelude` and
`Vision.Image`:

```haskell
import Vision.Histogram
-- or
import qualified Vision.Histogram as H
```

## The Image type-class

Images implement the `Image` type-class.

This type-class give you, among other things, the `index` and the `shape`
methods to look up for pixel values and to get the image size.

## Manifest and Delayed images

To benefit from Haskell's purity, non-mutable images are represented in two
ways:

* the `Manifest` representation stores images in Haskell `Vector`s. `Grey`,
`HSV`, `RGB` and `RGBA` are *manifest* images ;

* the `Delayed` representation uses a function to produce image pixels. These
images can be efficiently chained to produce complex transformations. With some
inlining, Haskell compilers are able to produce fast algorithms by removing
intermediate structures when working with delayed images.
`GreyDelayed`, `HSVDelayed`, `RGBDelayed` and `RGBADelayed` are *delayed*
images.

The `convert` method from the `convertible` package can be used to convert
images between color-spaces and representations.

As most functions work with both representations and all color-spaces, you need
to help the type checker to choose the correct return type.

For example, if you want to convert an RGBA image to a greyscale image, use a
type annotation to inform the compiler the image type you want to get:

```haskell
toGrey :: RGBA -> Grey
toGrey = convert
```

When you only need to force the returned representation of a function, you can
use the `delayed` and `manifest` functions. These functions don't do anything
except enforcing types.

```haskell
makeMiniature = delayed . resize Bilinear (ix2 150 150)
```

See
[this file](https://github.com/RaphaelJ/friday-examples/blob/master/src/Delayed.hs)
for an example of a pipeline of delayed images.

## Masked images

Some images are not defined for each of their pixels. These are called masked
images and implement the `MaskedImage` type-class.

The `MaskedImage` type-class primarily gives you a `maskedIndex` method which
differs from `index` by its type :

```haskell
index       :: Image i       => i -> Point -> ImagePixel i
maskedIndex :: MaskedImage i => i -> Point -> Maybe (ImagePixel i)
```

Unlike `index`, `maskedIndex` doesn't always return a value. For convenience,
every `Image` instance is also a `MaskedImage` instance.

`DelayedMask` can be used to create a masked image.

## Create images from functions

Images are instance of the `FromFunction` type-class. This class provide a
method to create images from a function.

For example, if you want to create a black and white image from another image,
you could write:

```haskell
toBlackAndWhite :: (Image i, Convertible i Grey) => i -> Grey
toBlackAndWhite img =
    let grey = convert img :: Grey
    in fromFunction (shape img) $ \pt ->
            if grey `index` pt > 127 then 255
                                     else 0
```

However, most of the time, you will prefer to use the `map` function instead:

```haskell
toBlackAndWhite :: (Image i, Convertible i Grey) => i -> Grey
toBlackAndWhite img =
    let grey = convert img :: Grey
    in map (\pix -> if pix > 127 then 255 else 0) grey
```

## Load and save images from and to files

For modularity, storage operations are provided by separate packages.

The *[friday-devil](https://hackage.haskell.org/package/friday-devil)* package
provides FFI calls to the *DevIL* image library. The library is written in C and
supports a wide variety of image formats, including BMP, JPG, PNG, GIF, ICO and
PSD.

The popular *[JuicyPixels](https://hackage.haskell.org/package/JuicyPixels)*
package is a native Haskell library for decoding many image types.  Efficient,
zero copy, conversion between JuicyPixels and Friday Haskell types is provided
by
*[friday-juicypixels](http://hackage.haskell.org/package/friday-juicypixels)*.

# Examples

See the *[friday-examples](https://github.com/RaphaelJ/friday-examples)*
package for a set of examples.

# Benchmarks

The library has been sightly optimized and could be moderately used for
real-time applications.

The following graph shows how the library compares to two other image processing
libraries written in C. The fastest implementation for each algorithm is taken
as reference:

![Benchmark results](bench_results.png)

The main reason that performances are currently below OpenCV is that OpenCV
relies heavily on SIMD instructions.


changelog
-*-change-log-*-

v0.2.4.0 TDB
 * Minimum base is now 4.8

v0.2.3.1 December 2016
 * Merge fix for storable instance

v0.2.3.0 December 2016
 * Add Contour Tracing (Vision.Image.Contour)
 * Add pointer to friday-juicypixels
 * First release by TomMD - Raphael is MIA but it remains his package.

v0.2.2.0 June 2015
 * New 'computeP' function to compute an image in parallel.

v0.2.1.2 May 2015
 * Add test modules under `other-modules`.

v0.2.1.1 April 2015
 * Removes an orphan dependency to bytetring.
 * Fixes some warning when compiled with GHC 7.10.

v0.2.1.0 March 2015
 * Grey to HSV conversion.

v0.2.0.2 February 2015
 * Minor documentation improvements.

v0.2.0.1 January 2015
 * Setup.hs script.

v0.2 January 2015
 BREAKING CHANGES:
   * Origin is on the upper left corner of the image (was bottom left).
   * New simplified filters interface. The old interface is still used
     internally to create filters and is available in
     Vision.Image.Filter.Internal.
   * DevIL bindings (Vision.Image.Storage) are now in a separate package
     (friday-devil) and with an improved interface.
   * Benchmarks are now in a separate package (friday-bench).
   * Examples are now in a separate package (friday-examples).
   * Pixel instance is no more required for MaskedImage and Image instances.

 * SCW thresholding.
 * New types aliases for filters.
 * New (!) and (!?) operators for image and histogram indexing.
 * Classes and image types are in two different files (Vision.Image.Class and
   Vision.Image.Type).
 * NFData instance for Manifest images.

v0.1 August 2014
 * Initial release


bench/Benchmark.hs
{-# LANGUAGE BangPatterns, FlexibleContexts #-}
import Control.Monad.ST.Safe (ST)
import Criterion.Main
import Data.Int

import Vision.Image (
      Grey, HSV, RGBA, RGB, RGBDelayed, InterpolMethod
    )
import qualified Vision.Detector.Edge as D (canny)
import qualified Vision.Image as I
import Vision.Histogram (Histogram)
import qualified Vision.Histogram as H
import Vision.Primitive

path :: FilePath
path = "bench/image.jpg"

main :: IO ()
main = do
    Right io <- I.load Nothing path
    let !(Z :. h :. w) = I.shape rgb
        !halfSize      = Rect (w `quot` 2) (h `quot` 2)
                              (w `quot` 2) (h `quot` 2)
        !rgb           = I.convert io             :: RGB
        !rgba          = I.convert rgb            :: RGBA
        !grey          = I.convert rgb            :: Grey
        !edges         = canny' grey
        !hsv           = I.convert rgb            :: HSV
        !hist          = H.histogram Nothing grey :: H.Histogram DIM1 Int32
        !hist2D        = H.histogram2D (ix3 256 3 3) grey
                                                  :: H.Histogram DIM3 Int32

    defaultMain [
          bgroup "IO" [
              bench "load" $ whnfIO $ I.load Nothing path
            ]
        , bgroup "conversion" [
              bench "RGB to grey"  $ whnf (I.convert :: RGB  -> Grey) rgb
            , bench "RGBA to grey" $ whnf (I.convert :: RGBA -> Grey) rgba
            , bench "RGBA to RGB"  $ whnf (I.convert :: RGBA -> RGB)  rgba
            , bench "RGB to RGBA"  $ whnf (I.convert :: RGB  -> RGBA) rgb
            , bench "RGB to HSV"   $ whnf (I.convert :: RGB  -> HSV)  rgb
            , bench "HSV to RGB"   $ whnf (I.convert :: HSV  -> RGB)  hsv
            ]
        , bgroup "crop" [
              bench "RGB" $ whnf (I.crop halfSize :: RGB -> RGB) rgb
            ]
        , bgroup "detector" [
              bench "Canny's edge detector" $ whnf canny' grey
            ]
        , bgroup "filter" [
              bench "erode"         $ whnf erode' grey
            , bench "blur"          $ whnf blur' grey
            , bench "gaussian blur" $ whnf gaussianBlur' grey
            , bench "scharr"        $ whnf scharr' grey
            , bench "sobel"         $ whnf sobel' grey
            , bench "mean"          $ whnf mean' grey
            ]
        , bgroup "flip" [
              bench "horizontal" $ whnf (I.horizontalFlip :: RGB -> RGB) rgb
            , bench "vertical"   $ whnf (I.verticalFlip   :: RGB -> RGB) rgb
            ]
        , bench "flood-fill" $ whnf floodFill' edges
        , bgroup "histogram" [
              bench "calculate 1D histogram of a grey image" $
                whnf (H.histogram Nothing :: Grey -> Histogram DIM1 Int32) grey
            , bench "calculate 3D histogram of a RGB image" $
                whnf (H.histogram Nothing :: RGB  -> Histogram DIM3 Int32) rgb
            , bench "calculate 3D histogram (9 regions) of a grey image" $
                whnf (H.histogram2D (ix3 256 3 3)
                                    :: Grey -> Histogram DIM3 Int32)
                     grey

            , bench "reduce an Int32 histogram" $ whnf H.reduce hist2D
            , bench "resize an Int32 histogram" $ whnf (H.resize (ix1 128))
                                                       hist

            , bench "cumulative Int32 histogram" $ whnf H.cumulative hist

            , bench "normalize histogram" $
                whnf (H.normalize 1
                        :: Histogram DIM1 Int32 -> Histogram DIM1 Double)
                     hist
            , bench "equalize grey image" $
                whnf (H.equalizeImage :: Grey -> Grey) grey

            , bench "correlation comparison" $
                whnf (H.compareCorrel hist :: Histogram DIM1 Int32 -> Double)
                     hist
            , bench "chi-square comparison" $
                whnf (H.compareChi hist :: Histogram DIM1 Int32 -> Double) hist
            , bench "intersection comparison" $
                whnf (H.compareIntersect hist :: Histogram DIM1 Int32 -> Int32)
                     hist
            , bench "EMD comparison" $ whnf (H.compareEMD hist) hist

            , bench "2D correlation comparison" $
                whnf (H.compareCorrel hist2D :: Histogram DIM3 Int32 -> Double)
                     hist2D
            , bench "2D chi-square comparison 2D" $
                whnf (H.compareChi hist2D :: Histogram DIM3 Int32 -> Double)
                     hist2D
            , bench "2D intersection comparison 2D" $
                whnf (H.compareIntersect hist2D
                                            :: Histogram DIM3 Int32 -> Int32)
                     hist2D
            ]
        , bgroup "resize" [
              bench "truncate-integer 50%" $
                whnf (resize' I.TruncateInteger (ix2 (h `quot` 2) (w `quot` 2)))
                     rgb
            , bench "truncate-integer 200%" $
                whnf (resize' I.TruncateInteger (ix2 (h * 2) (w * 2))) rgb
            , bench "nearest-neighbor 50%" $
                whnf (resize' I.NearestNeighbor (ix2 (h `quot` 2) (w `quot` 2)))
                     rgb
            , bench "nearest-neighbor 200%" $
                whnf (resize' I.NearestNeighbor (ix2 (h * 2) (w * 2))) rgb
            , bench "bilinear 50%" $
                whnf (resize' I.Bilinear (ix2 (h `quot` 2) (w `quot` 2))) rgb
            , bench "bilinear 200%" $
                whnf (resize' I.Bilinear (ix2 (h * 2) (w * 2))) rgb
            ]
        , bgroup "threshold" [
              bench "simple threshold"   $ whnf threshold'         grey
            , bench "adaptive threshold" $ whnf adaptiveThreshold' grey
            , bench "Otsu's method"      $ whnf otsu'              grey
            , bench "SCW"                $ whnf scw'               grey
            ]


        , bgroup "application" [
              bench "miniature 150x150" $ whnf miniature rgb
            ]
        ]
  where
    canny' :: Grey -> Grey
    canny' !img = D.canny 2 256 1024 img

    erode' :: Grey -> Grey
    erode' !img = I.erode 1 `I.apply` img

    blur'  :: Grey -> Grey
    blur' !img =
        let filt = I.blur 1 :: I.Blur I.GreyPixel Int16 I.GreyPixel
        in filt `I.apply` img

    gaussianBlur' :: Grey -> Grey
    gaussianBlur' !img =
        let filt = I.gaussianBlur 1 Nothing :: I.Blur I.GreyPixel Float
                                                      I.GreyPixel
        in filt `I.apply` img

    sobel' :: Grey -> I.Manifest Int16
    sobel' !img = I.sobel 1 I.DerivativeX `I.apply` img

    scharr' :: Grey -> I.Manifest Int16
    scharr' !img = I.scharr I.DerivativeX `I.apply` img

    mean' :: Grey -> I.Manifest Float
    mean' !img =
        let filt = I.mean (ix2 3 3) :: I.Mean I.GreyPixel Int16 Float
        in filt `I.apply` img

    floodFill' :: Grey -> I.Grey
    floodFill' img =
        I.create $ do
            mut <- I.thaw img :: ST s (I.MutableManifest I.GreyPixel s)
            I.floodFill (ix2 5 5) 255 mut
            return mut

    resize' :: InterpolMethod -> Size -> RGB -> RGB
    resize' = I.resize

    threshold' :: Grey -> Grey
    threshold' !img = I.threshold (> 127) (I.BinaryThreshold 0 255) img

    adaptiveThreshold' :: Grey -> Grey
    adaptiveThreshold' !img =
        let filt :: I.AdaptiveThreshold I.GreyPixel Float I.GreyPixel
            filt = I.adaptiveThreshold (I.GaussianKernel Nothing) 1 0
                                       (I.BinaryThreshold 0 255)
        in filt `I.apply` img

    otsu' :: Grey -> Grey
    otsu' !img = I.otsu (I.BinaryThreshold 0 255) img

    scw' :: Grey -> Grey
    scw' !img = I.scw (ix2 5 5) (ix2 15 15) (5 :: Double)
                      (I.BinaryThreshold 255 0) img

    miniature !rgb =
        let Z :. h :. w = I.shape rgb
        in if w > h
              then resizeSquare $ I.crop (Rect ((w - h) `quot` 2) 0 h h) rgb
              else resizeSquare $ I.crop (Rect 0 ((h - w) `quot` 2) w w) rgb

    resizeSquare :: RGBDelayed -> RGB
    resizeSquare = I.resize I.Bilinear (Z :. 150 :. 150)
